Feature: Adds probability of improvement as an acquisition function

miguelgondu opened the pull request:
## Type of changes

- [x] New feature
- [x] Documentation / docstrings
- [x] Tests

## Checklist

- [x] I've formatted the new code by running `poetry run pre-commit run --all-files --show-diff-on-failure` before committing.
- [x] I've added tests for new code.
- [x]  I've added docstrings for the new code.

## Checklist for this PR in particular

- [x] Adding tests particular to the PI acquisition function.
- [x] Decide on whether we want to call these `UtilityFunction`s or `AcquisitionFunction`s. (We decided to keep them as `utility_functions`)
- [x] Decide on whether we want to write a whole new tutorial for these, or just expand the tutorial on BO. (We decided to add to an existing tutorial)
- [x] Add docstrings.

## Description

This PR expands the `decision_making` module by adding Probability of Improvement as a potential acquisition function. It refactors the tests that are joint to utility functions, and provides a tutorial for how to use Probability of Improvement in practice (pattern-matching from the current tutorial on Thompson Sampling).

Issue Number: N/A

Thomas-Christie commented on gpjax/decision_making/utility_functions/probability_of_improvement.py:
```diff
@@ -0,0 +1,110 @@
+        objective_posterior = posteriors[OBJECTIVE]
+        if not isinstance(objective_posterior, ConjugatePosterior):
+            raise ValueError(
+                "Objective posterior must be a ConjugatePosterior to draw an approximate sample."
```
I added this check to Thompson sampling due to utilising the approximate sampling functionality. We can at least update the error message here.

miguelgondu replied:
Sounds good, I'll update the error message.

Thomas-Christie commented on gpjax/decision_making/utility_functions/probability_of_improvement.py:
```diff
@@ -0,0 +1,110 @@
+                "Objective posterior must be a ConjugatePosterior to draw an approximate sample."
+            )
+
+        objective_dataset = datasets[OBJECTIVE]
```
Probably best to have something along the lines of 

```
if objective_dataset.X is None or objective_dataset.n == 0:
            raise ValueError("Objective dataset must contain at least one item")
```
given that we use the objective dataset to find `best_y`.

miguelgondu replied:
Addressed!

Thomas-Christie commented on gpjax/decision_making/utility_functions/probability_of_improvement.py:
```diff
@@ -0,0 +1,110 @@
+            predictive_dist = objective_posterior.predict(x_test, objective_dataset)
+
+            # Assuming that the goal is to minimize the objective function
+            best_y = objective_dataset.y.min()
```
It could be useful to define `best_y` as the minimum _posterior mean_ value at any of the observed points, which would also handle the case where observations could be noisy.

miguelgondu replied:
Sounds good. I've updated how we compute `best_y`. It is the first time I see it being computed like that, though. If I'm not mistaken, other frameworks use the min of the dataset.

Thomas-Christie replied:
Yeah - for instance this approach is mentioned in ["A benchmark of kriging-based infill criteria for noisy
optimization"](https://hal.science/hal-00658212/document), section 3.2.

Thomas-Christie commented on gpjax/decision_making/utility_functions/probability_of_improvement.py:
```diff
@@ -0,0 +1,110 @@
+            # Assuming that the goal is to minimize the objective function
+            best_y = objective_dataset.y.min()
+
+            return gaussian_cdf(
```
Instead of implementing our own `gaussian_cdf` function, IMO it makes sense to use the existing normal distribution provided by Tensorflow probability i.e. `normal = tfp.distributions.Normal(predicitve_dist.mean(), predictive_dist.stddev())` and use the cdf defined on this.

miguelgondu replied:
Another alternative would be to update `GaussianDistribution`, right? For now, I'll create a `normal` as you state.

Thomas-Christie commented on tests/test_decision_making/test_utility_functions/test_utility_functions.py:
```diff
@@ -0,0 +1,158 @@
+)
+
+
+@pytest.mark.filterwarnings(
```
Nice, this will help make testing additional utility functions a lot easier.

miguelgondu replied:
I'm glad!

Thomas-Christie commented on tests/test_decision_making/test_utility_functions/test_utility_functions.py:
```diff
@@ -0,0 +1,158 @@
+    posterior = generate_dummy_conjugate_posterior(dataset)
+    posteriors = {OBJECTIVE: posterior}
+    datasets = {OBJECTIVE: dataset}
+    ts_utility_builder = utility_function_builder(**utility_function_kwargs)
```
nit: Some of the names of variables in this file e.g. `ts_utility_builder` could do with updating now that these tests are more "utility function agnostic".

miguelgondu replied:
Solved!

Thomas-Christie reviewed (changes requested):
Nice job - have just left a few minor comments.

miguelgondu commented:
Hi @Thomas-Christie,

Thanks for the review! I've addressed all the comments. Could I ask you to take a second look? Let me know if anything else comes to mind. 

Cheers.

Thomas-Christie commented on gpjax/decision_making/utility_functions/probability_of_improvement.py:
```diff
@@ -0,0 +1,126 @@
+    More precisely, given a predictive posterior distribution of the objective
+    function $`f`$, the probability of improvement at a test point $`x`$ is defined as:
+    $$`\text{PI}(x) = \text{Prob}[f(x) < f(x_{\text{best}})]`$$
+    where $`x_{\text{best}}`$ is the minimizer of $`f`$ in the dataset.
```
Could update to something like `where $x_{\text{best}}$ is the minimiser of the posterior mean at previously observed values, to handle noisy observations`.

miguelgondu replied:
Updated! Forgot to change this.

Thomas-Christie commented on tests/test_decision_making/test_utility_functions/test_probability_of_improvement.py:
```diff
@@ -0,0 +1,68 @@
+    assert jnp.isclose(utility_values, expected_utility_values).all()
+
+
+if __name__ == "__main__":
```
nit: I think you forgot to remove this after debugging.

miguelgondu replied:
Absolutely. My bad.

miguelgondu commented:
All green!

thomaspinder commented:
Super work @miguelgondu ! Thanks for reviewing @Thomas-Christie
